chainntnfs/height_hint_cache.go

```diff
@@ -0,0 +1,301 @@
+		_, err := tx.CreateBucketIfNotExists(spendHintBucket)
+		if err != nil {
+			return err
+		}
+
+		_, err = tx.CreateBucketIfNotExists(confirmHintBucket)
+		return err
+	})
+}
+
+// CommitSpendHint commits a spend hint for the outpoints to the cache.
+func (c *heightHintCache) CommitSpendHint(heightHint uint32,
+	ops ...wire.OutPoint) error {
+
+	return c.db.Batch(func(tx *bolt.Tx) error {
+		spendHints := tx.Bucket(spendHintBucket)
+		if spendHints == nil {
+			return ErrCorruptedHeightHintCache
+		}
+
+		var hint bytes.Buffer
+		if err := channeldb.WriteElement(&hint, heightHint); err != nil {
+			return err
+		}
+
+		var outpoint bytes.Buffer
+		for _, op := range ops {
+			outpoint.Reset()
+			err := channeldb.WriteElement(&outpoint, op)
+			if err != nil {
+				return err
+			}
+
+			err = spendHints.Put(outpoint.Bytes(), hint.Bytes())
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// QuerySpendHint returns the latest spend hint for an outpoint. The bool
+// returned signifies whether we have a hint cached for this outpoint.
+func (c *heightHintCache) QuerySpendHint(op wire.OutPoint) (uint32, bool) {
+	var hint uint32
+	err := c.db.View(func(tx *bolt.Tx) error {
+		spendHints := tx.Bucket(spendHintBucket)
+		if spendHints == nil {
+			return ErrCorruptedHeightHintCache
+		}
+
+		var outpoint bytes.Buffer
+		if err := channeldb.WriteElement(&outpoint, op); err != nil {
+			return err
+		}
+
+		spendHint := spendHints.Get(outpoint.Bytes())
+		if spendHint == nil {
+			return ErrSpendHintNotFound
+		}
+
+		return channeldb.ReadElement(bytes.NewReader(spendHint), &hint)
+	})
+	if err != nil {
+		return 0, false
+	}
+
+	return hint, true
+}
+
+// PurgeSpendHint removes the spend hint for the outpoints from the cache.
+func (c *heightHintCache) PurgeSpendHint(ops ...wire.OutPoint) error {
+	return c.db.Batch(func(tx *bolt.Tx) error {
+		spendHints := tx.Bucket(spendHintBucket)
+		if spendHints == nil {
+			return ErrCorruptedHeightHintCache
+		}
+
+		var outpoint bytes.Buffer
+		for _, op := range ops {
+			outpoint.Reset()
+			err := channeldb.WriteElement(&outpoint, op)
+			if err != nil {
+				return err
+			}
+
+			err = spendHints.Delete(outpoint.Bytes())
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+}
+
+// CommitConfirmHint commits a confirm hint for the transactions to the cache.
+func (c *heightHintCache) CommitConfirmHint(heightHint uint32,
+	txids ...chainhash.Hash) error {
+
+	return c.db.Batch(func(tx *bolt.Tx) error {
+		confirmHints := tx.Bucket(confirmHintBucket)
+		if confirmHints == nil {
+			return ErrCorruptedHeightHintCache
+		}
+
+		var hint bytes.Buffer
+		if err := channeldb.WriteElement(&hint, heightHint); err != nil {
+			return err
+		}
+
+		var txHash bytes.Buffer
+		for _, txid := range txids {
+			txHash.Reset()
+			err := channeldb.WriteElement(&txHash, txid)
+			if err != nil {
+				return err
+			}
+
+			err = confirmHints.Put(txHash.Bytes(), hint.Bytes())
```

Comment:
I guess we should optimize by keeping the largest hint if it exists from before in the cache.

Reply:
Scratch that, guess that is added in a later commit.